[자동차 경주 구현] 동전 미션 제출합니다.

donghyun81 opened the pull request:
안녕하세요 리뷰어님! 리뷰 잘 부탁합니다!

궁금한 점이 있는데 
상수화과정에서 어디까지 상수화 과정을 거쳐야 하는지 궁금합니다. 예를 들어, 저희는 joinTo(＂,＂)에서 반점까지 상수화를 시켰는데 이 과정이 컨벤션 부합하는지 궁금합니다.

구현에 한계를 느껴서 매개변수 값을 mutable를 사용했는데
함수형 프로그래밍에서는 immutable을 사용해서 값의 변경을 최소화하는걸 권장하는데 현업에서는 어떻게 해결하는지 궁금합니다!

바쁜 시간 내주셔서 감사합니다.

추가 질문 사항: 
1. 익명을 구현하면서 생각보다 예외 케이스나 사용자 입장에서 고려할 부분이 많았습니다.
가볍게 생각하고 시작했는데 힘들거나 오래 걸리는 일이 생겼을 때 동료들과 어떻게 합의를 하는지 궁금합니다!

2. 안드로이드나 프론트 엔드와 같은 UX에 크게 관련이 있는 개발자들은 사용자의 편의를 위해서 개발을 해야 한다고 생각하는데,
제이든은 사용자 편의를 위해서 어떤 공부를 했었는지, 어떤 방법이 크게 도움이 되었는지 궁금합니다!

donghyun81 committed:
docs(README): 실행 순서에 따른 기능 목록 작성

donghyun81 committed:
feat(InputView): 경주할 자동차를 입력 받는 기능 구현

donghyun81 committed:
feat(InputView): 시도할 횟수 입력 기능 구현

donghyun81 committed:
feat(Car): 자동차 주행 기능 추가

donghyun81 committed:
feat(Racing): 자동차 이름과 횟수를 입력받는 메소드 추가

donghyun81 committed:
feat(Racing): 시도 횟수에 따른 자동차 이동 기능 추가

donghyun81 committed:
feat(OutputView): 자동차 상태 출력 기능 구현

donghyun81 committed:
feat(OutputView): 레이스 우승자 출력

donghyun81 committed:
feat: 레이스에 자동차 출력 기능 추가

donghyun81 committed:
feat(Racing): 레이스 결과 출력 추가

donghyun81 committed:
docs: 리팩토링 문서 추가

donghyun81 committed:
refactor: 최종 우승자 출력 역할 분리

- 우승자를 구하는 로직을 레이싱으로 이동

donghyun81 committed:
refactor: 하드코딩 상수화

donghyun81 committed:
refactor: 컨벤션 수정 및 하드코드 상수화

donghyun81 committed:
refactor(Racing): 내부 함수 접근 제한

donghyun81 committed:
refactor: 자동차 이동 기능 분리

donghyun81 committed:
feat(CarFactory): 자동차 생성하는 기능 추가

donghyun81 committed:
feat: 자동차 익명 이름들 추가 검증

donghyun81 committed:
refactor: 하드코드 상수화

donghyun81 committed:
feat: 중복된 이름 제거 기능 추가

donghyun81 committed:
refactor: 랜덤 값 생성 기능 분리

donghyun81 committed:
test(Car): 자동차 테스트

donghyun81 committed:
test(CarFactory): 익명 생성 테스트

donghyun81 committed:
feat(Application): 레이스 실행 추가

donghyun81 committed:
docs(README): 커밋 이력별 문서 수정

KwonDae commented on README.md:
```diff
@@ -1 +1,23 @@
+  - 하드코딩 상수화
+  - 내부 함수 접근 제한
+  - 자동차 이동 기능 분리
+  - 랜덤 값 생성 기능 분리
```
README를 커밋 단위와 일치하도록 잘 작성해주셔서 가독성이 정말 좋았습니다 👍

KwonDae commented on src/main/kotlin/Car.kt:
```diff
@@ -0,0 +1,12 @@
+data class Car(val name: String) {
+    init {
+        require(name.length < CAR_NAME_LENGTH) { CAR_NAME_ERROR }
```
require 활용 👍

KwonDae commented on src/main/kotlin/InputView.kt:
```diff
@@ -0,0 +1,15 @@
+import java.util.Scanner
+
+class InputView {
```
InputView와 OutputView를 잘 분리해주셨네요 👍 
그렇다면 InputView에서도 입력값에 대해 스스로 검증할 수 있도록 역할을 부여해보는건 어떨까요?

donghyun81 replied:
디자인 패턴을 공부할때는 View 파일에서 따로 validate를 하면 View의 의미가 퇴색된다고 생각해서 넣지 않았습니다.
디자인 패턴에서 View의 역할을 생각하지 않았을 때는 사용자의 입력을 읽은 함수를 반환할 때 필요한 검증하는 게 좋다는 생각이 들어서 View에서 검증하는 역할을 부여하는 게 좋을 것 같아서 수정하겠습니다.

KwonDae commented on src/test/kotlin/CarTest.kt:
```diff
@@ -0,0 +1,23 @@
+        assertThat(actual).isEqualTo(car.position)
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = ["abcdef", "익명123", "12345"])
+    fun `자동차 이름 테스트`(name: String) {
+        assertThrows<IllegalArgumentException> {
+            Car(name)
+        }
+    }
```
> 각 자동차에 이름을 부여할 수 있다. 자동차 이름은 5자를 초과할 수 없다.

5글자를 초과해야 에러가 발생하는 요구조건 이겠죠?

donghyun81 replied:
테스트 함수명에 관한 설명이 부족했던 것 같습니다! 수정하겠습니다!

KwonDae commented on src/main/kotlin/Const.kt:
```diff
@@ -0,0 +1,17 @@
+const val INIT_POSITION = 0
+const val TRY_NUMBER_UNTIL = 10
+const val MOVE_MIN_NUMBER = 4
+const val CAR_NAME_LENGTH = 5
+const val ANONYMITY_MIN_COUNT = 1
+const val ANONYMITY = "익명"
+const val INPUT_CAR_NAME = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분)."
+const val INPUT_ATTEMPTS = "시도할 횟수는 몇 회인가요?"
+const val RESULT = "실행결과"
+const val FINAL_WINNER = "최종 우승자:"
+const val CAR_NAME_ERROR = "[ERROR]: 자동차 이름은 5자를 내외입니다."
+const val ATTEMPT_INPUT_ERROR = "[ERROR]: 시도 횟수를 입력해주세요."
+const val CAR_POSITION = "-"
+const val CAR_DELIMITER = ","
+const val WINNER_DELIMITER = ", "
+const val CAR_STATE_DELIMITER = " : "
+const val CAR_DUPLICATE_REMOVE = "중복된 차이름을 제거하였습니다."
```
상수를 한곳에 모아두셨네요..!
현재는 프로젝트 크기가 작아 이렇게 관리할 수 있겠지만,
만약 프로젝트에서 사용하는 상수가 1000개 이상이 된다면 어떤 일이 발생할까요?

KwonDae replied:
> 상수화과정에서 어디까지 상수화 과정을 거쳐야 하는지 궁금합니다. 예를 들어, 저희는 joinTo(＂,＂)에서 반점까지 상수화를 시켰는데 이 과정이 컨벤션 부합하는지 궁금합니다.

사실 이거는 정답이 없다고 생각합니다.
요구사항에 대해 최대한 공통적인 부분을 상수로 놓아 중복을 줄이게 할 수 있고
프로젝트마다 컨벤션을 정하기 나름이기에 페어의 의견에 따라 얼마든지 바뀔 수 있는 부분이라 생각해요.

donghyun81 replied:
자동차 경주 미션에서는 상수화 할 코드가 적어 하나의 파일에 모아 두었습니다. 
하지만, 상수가 1000개 이상이 된다면 오히려 가독성이 더 떨어지는 거 같아 
각 클래스 별로 관리 또는 상수 파일을 사용하는 것을 고민해보면 더 좋을 것 같습니다. 
혹시 현업에서는 어떻게 관리하는지 궁금합니다!

KwonDae commented on src/main/kotlin/CarFactory.kt:
```diff
@@ -0,0 +1,14 @@
+class CarFactory {
+    fun createCar(carNames: List<String>): List<Car> {
+        val anonymousCounts = (ANONYMITY_MIN_COUNT..carNames.count { it.isBlank() }).toMutableList()
+        return carNames.map { carName -> getAnonymousCar(carName, anonymousCounts) }
+    }
+
+    private fun getAnonymousCar(
+        name: String,
+        anonymousCounts: MutableList<Int>,
+    ): Car {
+        if (name.isBlank()) return Car(ANONYMITY + anonymousCounts.removeFirst())
+        return Car(name)
+    }
+}
```
익명 생성은 어떤 요구사항으로 생긴 로직일까요?

donghyun81 replied:
제가 앱의 사용자로서 자동차 경주를 한다면,
매번 경주할 때마다 이름을 설정하는 것이 어렵게 느껴져서,
이름을 설정하지 않아도 자동으로 부여되도록 구현하는 게 더 좋다고 느껴져서 구현했습니다.
프리코스를 진행하면서도 사용자 입장으로 생각해서 추가로 구현했던 부분들이 있었는데,
현업에서는 이런 생각이나 의견이 어떻게 반영되는지 궁금합니다!

KwonDae commented on src/main/kotlin/Racing.kt:
```diff
@@ -0,0 +1,47 @@
+    ) {
+        outputView.printResultTitle()
+        repeat(attempts) {
+            carMove(cars)
```
car에 메시지를 던지는 방식 👍

KwonDae commented on src/main/kotlin/RandomGenerator.kt:
```diff
@@ -0,0 +1,5 @@
+import kotlin.random.Random
+
+class RandomGenerator {
+    fun generate() = Random.nextInt(TRY_NUMBER_UNTIL)
+}
```
이부분은 조금 어려울 수 있으니 생각만 해보고 넘어가도 좋습니다.
현재는 Random에 의존하고 있기에 RandomGenerator 를 활용해 테스트를 할 수 없었습니다
(CarTest의 `자동차 이동 테스트`)

어떻게 하면 Car를 테스트 가능하도록 할 수 있을까요?
[전략패턴](https://victorydntmd.tistory.com/292)을 참고해보셔도 좋습니다 !

donghyun81 replied:
조언 감사합니다!
조금 더 공부해서 적용해보고 리뷰요청에서 궁금한 점을 더 여쭤 보겠습니다!

KwonDae commented on src/main/kotlin/CarFactory.kt:
```diff
@@ -0,0 +1,14 @@
+class CarFactory {
+    fun createCar(carNames: List<String>): List<Car> {
+        val anonymousCounts = (ANONYMITY_MIN_COUNT..carNames.count { it.isBlank() }).toMutableList()
```
> 함수형 프로그래밍에서는 immutable을 사용해서 값의 변경을 최소화하는걸 권장하는데 현업에서는 어떻게 해결하는지 궁금합니다!

현업에서도 물론 mutable을 사용하기도 합니다. 다만, 최대한 함수형 프로그래밍에서의 불변성을 이용하여 외부에서 올 수 있는 다양한 오류로부터 자유로워지게 하기 위해 노력하고 있습니다.

KwonDae reviewed (changes requested):
안녕하세요 동전!
이번 레벨1에서 첫 미션에서 함께할 제이든입니다 :)

첫 미션 잘 진행해주셨어요. 함께 고민해보면 좋을 만한 부분에 코멘트를 남겼으니
천천히 확인해주시고, 궁금한 부분이 있다면 언제든 DM 주세요!

donghyun81 committed:
docs: 피드백을 반영해서 수정사항 추가

donghyun81 committed:
feat: 전략 패턴을 이용해서 자동차 이동을 테스트 할 수 있도록 수정

donghyun81 committed:
test(CarTest)): 자동차가 4이상일 경우 위치를 상승시키는 테스트로 변경

donghyun81 committed:
test(CarTest)): 자동차가 4이상일 경우 위치를 상승시키는 테스트로 변경

donghyun81 committed:
refactor: 사용자에게 데이터를 읽는 함수에 예외처리 이동

donghyun81 committed:
Update CarTest.kt

donghyun81 committed:
test(CarFactoryTest): 자동차 익명 생성 성공 및 실패 케이스로 변경

donghyun81 committed:
test(Car): 함수명 세부표현으로 수정

donghyun81 committed:
refactor: 사용하는 위치에 상수 이동

donghyun81 committed:
refactor: 추출 숫자가 비었을 때 출력하는 예외 메시지 상수화

donghyun81 committed:
feat(InputView): 자동차 이름이 5자 이상일 경우 예외처리 추가

donghyun81 committed:
refactor: 숫자에서 이동 시도 숫자로 클래스명 변경

donghyun81 committed:
refactor(CarService): Racing 역할을 CarService로 분리

donghyun81 committed:
feat: 익명 뒤에 붙는 숫자를 랜덤으로 추가하도록 수정

- 익명1, 익명32를 사용자가 임의로 추가 할 경우 순서의 의미가 사라지고
중복된 순서가 생겼을때 프로그램 임의로 익명을 제거 할 수 있어서 구현 변경

donghyun81 committed:
refactor: 역할을 고려해서 이름 변경

donghyun81 committed:
feat: 5자 이내의 이름은 통과하도록 수정

donghyun81 committed:
test: 숫자의 값은 같지만 타입이 다를 경우에 제거되는지 테스트

donghyun81 committed:
feat: 소수점이 포함된 익명은 toInt를 통해서 수정되지 않도록 변경

donghyun81 committed:
feat(CarService): 자동차 객체를 포지션을 통해 비교 할 수 있도록 변경

donghyun81 committed:
test(CarService): 위치가 제일 높은 자동차들을 반환하는 테스트 추가

donghyun81 committed:
test(CarFactory): 각 테스트에 사용할 변수를 필드로 이동

donghyun81 committed:
debug(InputView): 입력을 두번 받는 오류 수정

donghyun81 committed:
docs: 기능 추가 및 변경사항 작성

donghyun81 committed:
refactor: 숫자 생성기들 의존성을 줄이기 위해 생성자 프로퍼티로 이동

KwonDae commented on README.md:
```diff
@@ -14,10 +14,25 @@
   - 자동차 이동 기능 분리
-  - 랜덤 값 생성 기능 분리
\ No newline at end of file
+  - 랜덤 값 생성 기능 분리
+
+- 수정사항
+  - 랜덤 숫자 생성기를 포함한 자동차 테스트로 수정
+    - 랜덤 숫자 생성기를 통해 생성한 값을 자동차 이동에 추가하기 위한 전략 패턴 활용
+  - InputView에서 스스로 검증할 수 있도록 역할 부여
+  - 테스트 함수명 수정
+  - 상수 분리
+  - 익명 생성에 관한 고민
```
적어주신 수정사항 덕분에 확인하기가 정말 좋았어요 🙌

KwonDae commented on src/main/kotlin/InputView.kt:
```diff
@@ -3,13 +3,26 @@ import java.util.Scanner
+        return requireNotNull(scanner.nextLine().toIntOrNull()) { ATTEMPT_INPUT_ERROR }
+    }
+
+    companion object {
+        private const val CAR_NAME_ERROR = "[ERROR]: 자동차 이름은 5자를 내외입니다."
+        private const val INPUT_CAR_NAME = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분)."
+        private const val INPUT_ATTEMPTS = "시도할 횟수는 몇 회인가요?"
+        private const val ATTEMPT_INPUT_ERROR = "[ERROR]: 시도 횟수를 입력해주세요."
+        private const val CAR_NAME_LENGTH = 5
+        private const val CAR_DELIMITER = ","
```
> 자동차 경주 미션에서는 상수화 할 코드가 적어 하나의 파일에 모아 두었습니다.
> 하지만, 상수가 1000개 이상이 된다면 오히려 가독성이 더 떨어지는 거 같아
> 각 클래스 별로 관리 또는 상수 파일을 사용하는 것을 고민해보면 더 좋을 것 같습니다.
> 혹시 현업에서는 어떻게 관리하는지 궁금합니다!

회사 또는 팀마다 다를 수 있지만,
공통으로 사용되는 상수가 아니라면 범위 지정자를 두는게 관리하는편입니다!

KwonDae commented on src/test/kotlin/CarTest.kt:
```diff
@@ -1,23 +1,38 @@
     @ParameterizedTest
-    @ValueSource(strings = ["abcdef", "익명123", "12345"])
-    fun `자동차 이름 테스트`(name: String) {
+    @ValueSource(strings = ["abcdef", "익명1234", "123456"])
+    fun `자동차 이름이 5자 초과일 경우 예외 발생 테스트`(name: String) {
```
"abcd, jason" 을 실행해도 에러가 나는 이유는 무엇일까요?

KwonDae commented on src/main/kotlin/CarFactory.kt:
```diff
@@ -1,14 +1,28 @@
+        return cars.distinct()
+    }
+
+    private fun initAnonymousNumbers(carsName: List<String>): List<Int> {
+        return carsName.filter { carName -> carName.startsWith(ANONYMITY) }
+            .mapNotNull { anonymousName -> anonymousName.removePrefix(ANONYMITY).toOnlyIntOrNull() }
```
> 제가 앱의 사용자로서 자동차 경주를 한다면,
> 매번 경주할 때마다 이름을 설정하는 것이 어렵게 느껴져서,
> 이름을 설정하지 않아도 자동으로 부여되도록 구현하는 게 더 좋다고 느껴져서 구현했습니다.
> 프리코스를 진행하면서도 사용자 입장으로 생각해서 추가로 구현했던 부분들이 있었는데,
> 현업에서는 이런 생각이나 의견이 어떻게 반영되는지 궁금합니다!

좋은 개발 방향이라고 생각합니다.
개발자는 개인적으로 요구사항에 대해서만 구현하는 것 외에도 사용자 입장에서 UX 친화적으로 개발하는 것 또한 능력이라고 생각해요.


다만, 조금 어려웠던 부분은 과연 빈칸으로 사용자가 입력했을 때 디폴트로 넣어주는 이름은 어디서 넣어주고 있는지 찾기가 어려웠어요.
사용자가 이름을 입력하든 하지 않든 위 함수를 거쳐가게 되어있는데, 과연 동료 개발자가 이 함수가 두가지 기능을 모두 할 것이라 예상하기 쉬울까요?

```kotlin
// in Controller
val inputNames = inputView.getNames() ?: defaultNames()
```

이런식으로 가독성을 고려해보면 어떨까요?

KwonDae commented on src/test/kotlin/CarTest.kt:
```diff
@@ -1,23 +1,38 @@
+        tryNumbers: List<Int>,
+        actual: Int,
+    ) {
+        val car = Car("동전", tryMoveNumberStrategy = TryMoveNumbersExtractor(tryNumbers))
```
이제 자동차 움직임을 테스트 할 수 있게 되었네요 👍

KwonDae reviewed (approved):
1단계 피드백 반영해주시느라 고생많으셨습니다.
2단계에서는 Controller의 역할을 배우게 되실텐데

현재의 구조에서 변경해야 할 부분들이 많을거에요.
차근차근 각자 역할에 맞게 분배해보시면서 남겨드린 피드백 참고하여 끝까지 파이팅해주세요! :)
